content/languages/python/authoring/index.md

```diff
@@ -0,0 +1,213 @@
+
+### Reference solution
+
+If the test suite happens to use a reference solution to calculate expected values (what [should be avoided](/authoring/guidelines/submission-tests/#reference-solution), but is not always possible), or some kind of reference data, precalculated arrays, etc., it can be redefined, overwritten, or accessed directly by user solution. To avoid this, it should be defined in a non-global scope, local to the testing function, `it` block, or `describe` block. The reference solution or data _must not_ be defined in [Preloaded code](/authoring/guidelines/preloaded/).
+
+### Calling assertions
+
+Python testing framework provides a set of useful [assertions](/languages/python/codewars-test/#assertions), but when used incorrectly, they can cause a series of problems:
+- Stacktraces of a crashing user solution can reveal details which should not be visible,
+- Reference soluton and user solution called in wrong order can mutate input in undesirable way allowing for abuse or leading to bugs,
+- Use of an assertion not suitable for given case can cause incorrect test results,
+- Incorrectly used assertion can produce confusing or unhelpful messages.
+
+To avoid above problems, calls to assertion functions should respect some rules:
+- Expected and actual values should be calculated _before_ a call to assertion method. Parameters passed to the assertion should not be function call expressions, but values calculated directly before.
+- If a reference solution is used to calculate expected value, it _must_ be either called _before_ user solution, or receive a (deep) copy of input data.
+- Assertion message _must not_ be composed from a mutable data which could be potentially modified by a user or reference soltion.  
+- Appropriate assertion function should be used for a given test. `assert_equals` is not suitable in all situations. Use `assert_approx_equals` for floating point comparisons, `expect` for tests on boolean values, `expect_error` to test error handling.
+- Some additional attention should be paid to the order of parameters passed to assertion functions. It differs between various assertion libraries, and it happens to be quite often confused by authors, mixing up `actual` and `expected` in assertion messages. For Python Testing framework, the order is `(actual, expected)`.
+- One somewhat distinctive feature of Python assertions is that by default, a failed assertion does not cause a test case to fail early. It can lead to unexpected crashes when an actual value had already been asserted to be invalid, but execution of the current test case was not stopped and following assertions continue to refer to it. This behavior can be ovrriden by passing `allow_raise=True` argument to the assertion functions which support it.
```

Comment:
```suggestion
- One somewhat distinctive feature of Python assertions is that by default, a failed assertion does not cause a test case to fail early. It can lead to unexpected crashes when an actual value had already been asserted to be invalid, but the execution of the current test case was not stopped and following assertions continue to refer to it. This behavior can be overridden by passing the `allow_raise=True` argument to the assertion functions which support it.
```